Flow router integration

PhilippSpo opened the pull request:
Started working on the flow-router integration. It is not perfect yet, but as far as I (and some others I asked to test) can tell everything is working fine.
One big concern I have is that I completely ignored the `ensureSingedIn` method and used the one that @arunoda shows in the flow-router [docs](https://github.com/meteorhacks/flow-router#middlewares), which seems to do the trick just fine..
One more thing that is not too beautiful right now is that you have to define the FlowLayout and the region where the useraccounts template gets rendered for each route like this:
````javascript
var stdConfig = {
  layoutTemplate: "masterLayout",
  renderLayout: {
      nav: "nav",
      footer: "footer"
  },
  contentRange: "area"
};
AccountsTemplates.configureRoute('signIn', stdConfig);
````
But I think it is an easy one to make that a global config variable on the AT object.
I deployed the materialize version of useraccounts using FlowRouter [here](http://useraccounts-materialize-flow.meteor.com/)
The Demo code is [here](https://github.com/PhilippSpo/useraccounts-materialize-flow).
There is one strange thing still happening: The `postSubmitRedirect` is not working on verifyEmail. haven't looked into this yet.

splendido commented:
Thanks @PhilippSpo, that's indeed a good start!

@arunoda could you please have a look at this, just to confirm it's spinned off the right way?

Thanks!

arunoda commented on lib/core.js:
```diff
@@ -652,57 +654,48 @@ AT.prototype.setupRoutes = function() {
+                    }],
+                    action: function(params) {
+                        renderLayout[contentRange] = template;
+                        FlowLayout.render(layoutTemplate, renderLayout);
```
Hope `renderLayout` contains some data for the layout?

PhilippSpo replied:
renderLayout needs to be defined per route like so:
````javascript
renderLayout: {
    nav: "nav",
    footer: "footer"
}
````
and is {} if there is nothing defined.

arunoda commented on package.js:
```diff
@@ -28,7 +30,9 @@ Package.on_use(function(api) {
         "softwarerero:accounts-t9n@1.0.6",
-        "iron:router@1.0.7",
+        "meteorhacks:flow-router@1.1.1",
+        "meteorhacks:flow-layout@1.0.1",
```
try updating flow layout to `1.0.2`. We fixed a major bug which affects when switching between layouts.

splendido commented:
Thanks @arunoda for the quick review!

Let me know if we should merge this and continue the discussion on #308

arunoda commented:
@rclai I think fast render has been used  for user status check on the middleware. In flow router, there is no way to wait in the router level. Need to do it in the layout layer. 

I think that's the reason for using fast render. May be we can do something better.

PhilippSpo commented:
@arunoda is right. That's the reason why I used fast render. If there is a better way go for it :)

splendido commented:
So, is there any way to keep the current url in the address bar while displaying a login form for protected rotues?

PhilippSpo commented:
Yes this can be done with FlowLayout. I'll look into this later.

PhilippSpo commented:
The current url stays in the address bar now, since the routes don't change. Once the process is complete (successfully signed in) the route gets rerun so the `ensureSignedIn` function gets rerun.
We don't depend on FastRender anymore now.
I also renamed some of my variables to be more self-explanatory and also added default variables for them as with `defaultLayout`.

splendido commented:
Thank you Philipp!

splendido commented:
I guess the only way for now to use this it to clone `useraccounts:core` locally and switch to `flow-router-integration` branch by hands...

Hopefully I'll have time to work on v2.0 over the week-end!

splendido commented:
...what about having some *documentation update*? ;-)

PhilippSpo commented:
Sure. As a PR on master ?

PhilippSpo commented:
This is how I do it right now
https://github.com/PhilippSpo/useraccounts-materialize-flow/blob/master/README.md

splendido commented:
I'd say yes, you can add a section to let people know there's something in the air.
Having some instruction about how to test it (local clone, etc..) would be great!

I truly appreciate your contribution, while at the same time I'm sorry to ask you so much...

PhilippSpo commented:
No problem :)
Want to have it in Guide or Readme?

rclai commented:
What great timing! I was just getting into building an app with flow and needed useraccounts to work with it. Thanks.

splendido commented:
@rclai let we know how it goes! ;-)

rclai commented:
Hey @PhilippSpo, so I noticed that if you attach `AccountsTemplates.ensureSignedIn` globally like this:

```js
FlowRouter.middleware(AccountsTemplates.ensureSignedIn);
```

The AccountsTemplates doesn't initialize properly and breaks. It only works when you set it as a middleware to specific routes.

PhilippSpo commented:
@rclai Yes I also noticed that, but I didn't get around to looking into that yet. I currently set it manually as middleware to every route, too.

rclai commented:
Okay.

benjamin79 commented:
Flow Router integration breaks with newest version of it. For example link to pw reset leads to login page. Not sure if the deprecation of middlewares for triggers is the reason.

arunoda commented:
I'd happy to help.
No you can use middlewares for now and it'll work.
We had an issue with invoking "FlowRouter.go('/')", but we fixed it with
`1.14.1`

On Wed, Jun 10, 2015 at 2:42 AM benjamin79 <notifications@github.com> wrote:

> Flow Router integration breaks with newest version of it. For example link
> to pw reset leads to login page. Not sure if the deprecation of middlewares
> for triggers is the reason.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/meteor-useraccounts/core/pull/327#issuecomment-110505648>
> .
>

PhilippSpo commented:
Thats interesting... It only happens when you go to the forgot-password page and reload the page. Then the `setupRoutes` method gets called after the error  "There is no route for the path: /forgot-password"...

PhilippSpo commented:
![image](https://cloud.githubusercontent.com/assets/8394418/8071562/ae1b9e8c-0f0b-11e5-9110-5eafeb6f0c48.png)
@arunoda it is a problem that the useraccounts routes get setup once Meteor.startup fires, right?

arunoda commented:
I think so. SignIn route navigation call fired before the route setup.

On Wed, Jun 10, 2015 at 4:29 AM Philipp Sporrer <notifications@github.com>
wrote:

> [image: image]
> <https://cloud.githubusercontent.com/assets/8394418/8071562/ae1b9e8c-0f0b-11e5-9110-5eafeb6f0c48.png>
> @arunoda <https://github.com/arunoda> it is a problem that the
> useraccounts routes get setup once Meteor.startup fires, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/meteor-useraccounts/core/pull/327#issuecomment-110527579>
> .
>

PhilippSpo commented:
I think that this worked with an earlier version of FlowRouter..

PhilippSpo commented:
Yup it is working with `meteorhacks:flow-router@=1.9.0` for example.

splendido commented:
@PhilippSpo, could you please try to get rid of this two-step route definition and call `setupRoute` directly within `configureRoute`?

PhilippSpo commented:
@splendido I will. I just want to point out that this is how it is in the iron-router implementation (https://github.com/meteor-useraccounts/core/blob/master/lib/client.js#L454), where it maybe should also change..

PhilippSpo commented:
#428

splendido commented:
Yes! That should be changed too, I'll try to do it for v2.0....
See also [a couple of posts from here](https://github.com/meteor-useraccounts/core/issues/308#issuecomment-107772985)
